vm/devices/storage/disklayer_sqlite/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+        true
+    }
+
+    async fn read(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+        mut marker: SectorMarker<'_>,
+    ) -> Result<(), DiskError> {
+        let sector_count = (buffers.len() / self.meta.sector_size as usize) as u64;
+        let end_sector = sector + sector_count;
+        tracing::trace!(sector, sector_count, "read");
+        if end_sector > self.meta.sector_count {
+            return Err(DiskError::IllegalBlock);
+        }
+
+        let valid_sectors = unblock({
+            let conn = self.conn.clone();
+            let end_sector = sector + (buffers.len() as u64 / self.meta.sector_size as u64);
+            let sector_size = self.meta.sector_size;
+            move || read_sectors(conn, sector_size, sector, end_sector)
+        })
+        .await
+        .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        for (s, data) in valid_sectors {
+            let offset = (s - sector) as usize * self.meta.sector_size as usize;
+            let subrange = buffers.subrange(offset, self.meta.sector_size as usize);
+            let mut writer = subrange.writer();
+            match data {
+                SectorKind::AllZero => writer.zero(self.meta.sector_size as usize)?,
+                SectorKind::AllOne => writer.fill(0xff, self.meta.sector_size as usize)?,
+                SectorKind::Data(data) => writer.write(&data)?,
+            };
+
+            marker.set(s);
+        }
+
+        Ok(())
+    }
+
+    async fn write(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+        _fua: bool,
+    ) -> Result<(), DiskError> {
+        self.write_maybe_overwrite(buffers, sector, true).await
+    }
+
+    fn write_no_overwrite(&self) -> Option<impl WriteNoOverwrite> {
+        Some(self)
+    }
+
+    async fn sync_cache(&self) -> Result<(), DiskError> {
+        tracing::trace!("sync_cache");
+
+        (self.conn.lock())
+            .pragma_update(None, "wal_checkpoint", "FULL")
+            .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        Ok(())
+    }
+
+    async fn unmap(
+        &self,
+        sector_offset: u64,
+        sector_count: u64,
+        _block_level_only: bool,
+        next_is_zero: bool,
+    ) -> Result<(), DiskError> {
+        tracing::trace!(sector_offset, sector_count, "unmap");
+        if sector_offset + sector_count > self.meta.sector_count {
+            return Err(DiskError::IllegalBlock);
+        }
+
+        unblock({
+            let conn = self.conn.clone();
+            move || unmap_sectors(conn, sector_offset, sector_count, next_is_zero)
+        })
+        .await
+        .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        Ok(())
+    }
+
+    fn unmap_behavior(&self) -> UnmapBehavior {
+        // This layer zeroes if the lower layer is zero, but otherwise does
```

Comment:
Is this true? It seems like unmap will effectively zero everything.

Reply:
Sounds like the actual `unmap_sectors` semantics, as written, are correct though?

If so, I'll swap this over to `UnmapBehavior::Zeroes`

Reply:
Yeah